### DISCUSSION TITLE
Flet + FastAPI thoughts and example

### DISCUSSION THREAD

**Discussion by flokapi** [flet-dev/flet#1883] [Ideas]
Hi, i checked out the new feature on how to to run serve Flet Apps using FastAPI and i must say it's really cool.

Here is a simple demo project with some more thoughts: https://github.com/flokapi/flet-fastapi-example

Key points:

- I like it that it's possible to implement and run both the front end (Flet) and the back end (FastAPI) in the same project. It's still possible to clearly separate both of them.
- Using this approach, the Flet code is running on the server and this has some benefits because the Flet App can communicate directly with FastAPI using function calls and without going through HTTP requests (In the past i tried to make HTTP requests from Pyodide but without success)
- I propose a way to register the API endpoint functions and to pass them over as a dictionary to the Flet App, which enables the Flet App to access them while being defined in  sub-package
- There is also a example on how to define idle functions both in the API and in for each instance of Flet App (opened window). All of this only using async (no separate thread or child process).
- There is an example on how to publish this app as a Docker container locally or on a server using a new tool called Dockup

**Comment by FeodorFitsner** [comment]
Lovely!

Could you please explain what's the purpose of "idle" function in FastAPI?

I was so excited when implemented Flet into FastAPI that immediately started thinking about making a "dashboard" example for showing some stats from an API.

Do you think we should extend [flet.dev article](https://flet.dev/docs/guides/python/deploying-web-app/running-flet-with-fastapi) with your example/notes or you could create a separate blog post? I feel like we should cover more of Flet+FastAPI :)

_Flet - a better UI for FastAPI_ 🚀

**Reply by flokapi** [reply]
> Could you please explain what's the purpose of "idle" function in FastAPI?

An idle function in FastAPI could be used for recurring tasks such as web scrapping and update a database.

> Do you think we should extend [flet.dev article](https://flet.dev/docs/guides/python/deploying-web-app/running-flet-with-fastapi) with your example/notes or you could create a separate blog post? I feel like we should cover more of Flet+FastAPI :)

I don't have a blog at the moment and you are welcome to reuse the example and notes on your website. I can provide a review if you like.
I added a branch called "without-dockup" to the repository which is more simple to apprehend.
Maybe once Dockup is more mature and the `proxy_path` parameter is available in the package it makes sense to share it more broadly.

> I was so excited when implemented Flet into FastAPI that immediately started thinking about making a "dashboard" example for showing some stats from an API.

There is also another branch with a super basic API request counter.

The code could be simplified a bit further by removing the idle function in gui.py and using directly `await asyncio.create_task(counter.idle())`

Also, i noticed that lambda functions do not support async, therefore i could not use it to directly call the callback function as an object attribute and i used and intermediary function. I wonder whether there is a different way of doing this.
```
async def minus_click(e):
    await self.minus()
```

**Reply by flokapi** [reply]
Here is a blog post proposal: https://github.com/flokapi/flet-fastapi-example/blob/main/blog_post.md
